File: javascript/bridgeTorch/javascript/bridgeTorch.js
(function() {
    function Person(id, timeToCrossMinutes) {
        this.id = id;
        this.timeToCrossMinutes = timeToCrossMinutes;
    }


    var TorchSide = {
        LEFT: 'LEFT',
        RIGHT: 'RIGHT',
        flip: function(torchSide) {
            if(this.LEFT === torchSide) {
                return this.RIGHT;
            } else if(this.RIGHT === torchSide) {
                return this.LEFT;
            } else {
                throw 'Unknown torchSide ' + torchSide;
            }
        }
    };


    function PersonsToCross(personsToCross) {
        this.personsToCross = personsToCross;
    }

    PersonsToCross.prototype.containsPerson = function(targetPerson) {
        for(var i = 0; i < this.personsToCross.length; i++) {
            var person = this.personsToCross[i];
            var personId = person.id;
            var targetPersonId = targetPerson.id;
            var matches = personId === targetPersonId;
            if(matches) {
                return true;
            }
        }

        return false;
    };

    PersonsToCross.prototype.timeToCrossMinutes = function () {
        var timesToCrossMinutes = this.personsToCross.map(function(person) { return person.timeToCrossMinutes });
        return timesToCrossMinutes.reduce(function (max, current) {
            if (current > max) {
                return current;
            } else {
                return max;
            }
        });
    };


    function State(leftSide, rightSide, torchSide, elapsedTimeMinutes) {
        this.leftSide = leftSide;
        this.rightSide = rightSide;
        this.torchSide = torchSide;
        this.elapsedTimeMinutes = elapsedTimeMinutes;
    }

    State.prototype.nextPossibleStates = function () {
        if(TorchSide.LEFT === this.torchSide) {
            return this.nextPossibleStatesFromLeft();
        } else if(TorchSide.RIGHT === this.torchSide) {
            return this.nextPossibleStatesFromRight();
        } else {
            throw 'Unknown torchSide ' + this.torchSide;
        }
    };

    State.prototype.nextPossibleStatesFromLeft = function() {
        var possiblePersonsToCross = this.possiblePersonsToCrossFromLeft();
        return possiblePersonsToCross.map(function (personsToCross) {
            return this.nextStateFromLeft(personsToCross);
        }, this);
    };

    State.prototype.nextPossibleStatesFromRight = function () {
        var possiblePersonsToCross = this.possiblePersonsToCrossFromRight();
        return possiblePersonsToCross.map(function (personsToCross) {
            return this.nextStateFromRight(personsToCross);
        }, this);
    };

    State.prototype.possiblePersonsToCrossFromLeft = function() {
        var possiblePersonsToCross = [];

        for(var i = 0; i < this.leftSide.length; i++) {
            for(var j = i + 1; j < this.leftSide.length; j++) {
                var person1 = this.leftSide[i];
                var person2 = this.leftSide[j];

                var personsToCross = new PersonsToCross([person1, person2]);

                possiblePersonsToCross.push(personsToCross);
            }
        }

        return possiblePersonsToCross;
    };

    State.prototype.possiblePersonsToCrossFromRight = function() {
        var possiblePersonsToCross = [];

        for(var i = 0; i < this.rightSide.length; i++) {
            var person = this.rightSide[i];
            var personsToCross = new PersonsToCross([person]);

            possiblePersonsToCross.push(personsToCross);
        }

        return possiblePersonsToCross;
    };

    State.prototype.nextStateFromLeft = function(personsToCross) {
        var nextLeftSide = this.leftSide.filter(function(person) {
            return !personsToCross.containsPerson(person);
        });

        var nextRightSide = this.rightSide.concat(personsToCross.personsToCross);

        var nextTorchSide = TorchSide.flip(this.torchSide);

        var nextElapsedTimeMinutes = this.elapsedTimeMinutes + personsToCross.timeToCrossMinutes();

        return new State(nextLeftSide, nextRightSide, nextTorchSide, nextElapsedTimeMinutes);
    };

    State.prototype.nextStateFromRight = function(personsToCross) {
        var nextLeftSide = this.leftSide.concat(personsToCross.personsToCross);

        var nextRightSide = this.rightSide.filter(function(person) {
            return !personsToCross.containsPerson(person);
        }, this);

        var nextTorchSide = TorchSide.flip(this.torchSide);

        var nextElapsedTimeMinutes = this.elapsedTimeMinutes + personsToCross.timeToCrossMinutes();

        return new State(nextLeftSide, nextRightSide, nextTorchSide, nextElapsedTimeMinutes);
    };

    State.prototype.isDone = function() {
        return this.leftSide.length === 0;
    };


    function States(states) {
        this.states = states;
    }

    States.prototype.nextStates = function() {
        var nextStatesArray = [];

        this.states.forEach(function(state) {
            var nextPossibleStates = state.nextPossibleStates();
            nextStatesArray = nextStatesArray.concat(nextPossibleStates);
        });

        return new States(nextStatesArray);
    };

    States.prototype.finalStates = function() {
        var currentStates = this;

        while(true) {
            var allDone = currentStates.isAllDone();
            if(allDone) {
                return currentStates;
            }

            currentStates = currentStates.nextStates();
        }
    };

    States.prototype.finalELapsedTimeMinutes = function() {
        var finalStates = this.finalStates();

        return finalStates.states.map(function (state) {
            return state.elapsedTimeMinutes;
        });
    };

    States.prototype.minFinalElapsedTimeMinutes = function() {
        var finalElapsedTimeMinutes = this.finalELapsedTimeMinutes();

        return finalElapsedTimeMinutes.reduce(function (min, current) {
            if (min > current) {
                return current;
            } else {
                return min;
            }
        });
    };

    States.prototype.isAllDone = function() {
        return this.states.every(function (state) {
            return state.isDone();
        });
    };


    var initialState = new State([new Person('A', 1), new Person('B', 2), new Person('C', 5), new Person('D', 8)], [], TorchSide.LEFT, 0);
    var initialStates = new States([initialState]);
    var minFinalElapsedTimeMinutes = initialStates.minFinalElapsedTimeMinutes();
    console.log(minFinalElapsedTimeMinutes)

})();

File: perl/bridgeTorch/bridge-torch.pl
#!/usr/bin/perl
use strict;
use warnings;


use Data::Dumper;
use List::Util qw(max min);

my $initialState = {
    leftSide => [
        {
            id => 'A',
            timeToCrossMinutes => 1
        }, {
            id => 'B',
            timeToCrossMinutes => 2
        }, {
            id => 'C',
            timeToCrossMinutes => 5
        }, {
            id => 'D',
            timeToCrossMinutes => 8
        }
    ],
    rightSide => [],
    torchSide => 'LEFT',
    elapsedTimeMinutes => 0
};

my $minTime = minTime([$initialState]);

print "$minTime\n";



sub nextPossibleStates {
    my $state = shift;

    my $torchSide = $state->{torchSide};
    if('LEFT' eq $torchSide) {
        return nextPossibleStatesFromLeft($state);
    } elsif('RIGHT' eq $torchSide) {
        return nextPossibleStatesFromRight($state);
    } else {
        die "Unknown torchSide $torchSide";
    }
}

sub nextPossibleStatesFromLeft {
    my $state = shift;

    my $possiblePersonsToCross = possiblePersonsToCrossFromLeft($state);

    my @nextStates = map {
        my $personsToCross = $_;

        nextStateFromLeft($state, $personsToCross);
    } @$possiblePersonsToCross;

    return \@nextStates;
}

sub nextPossibleStatesFromRight {
    my $state = shift;

    my $possiblePersonsToCross = possiblePersonsToCrossFromRight($state);

    my @nextStates = map {
        my $personsToCross = $_;

        nextStateFromRight($state, $personsToCross);
    } @$possiblePersonsToCross;

    return \@nextStates;
}

sub possiblePersonsToCrossFromLeft {
    my $state = shift;

    my $possiblePersonsToCross = [];

    for my $person1 (@{$state->{leftSide}}) {
        for my $person2 (@{$state->{leftSide}}) {
            my $person1Id = $person1->{id};
            my $person2Id = $person2->{id};

            if($person1Id eq $person2Id) {
                next;
            }

            my $personsToCross = [$person1, $person2];
            push(@$possiblePersonsToCross, $personsToCross);
        }
    }

    return $possiblePersonsToCross;
}

sub possiblePersonsToCrossFromRight {
    my $state = shift;

    my $possiblePersonsToCross = [];

    for my $person (@{$state->{rightSide}}) {
        my $personsToCross = [$person];
        push(@$possiblePersonsToCross, $personsToCross);
    }

    return $possiblePersonsToCross;
}

sub nextStateFromLeft {
    my ($state, $personsToCross) = @_;

    my $leftSide = $state->{leftSide};
    my $nextLeftSide = nextLeftSideFromLeft($leftSide, $personsToCross);
    my $rightSide = $state->{rightSide};
    my $nextRightSide = nextRightSideFromLeft($rightSide, $personsToCross);
    my $torchSide = $state->{torchSide};
    my $nextTorchSide = flipTorchSide($torchSide);
    my $elapsedTimeMinutes = $state->{elapsedTimeMinutes};
    my $nextElapsedTimeMinutes = nextElapsedTimeMinutes($elapsedTimeMinutes, $personsToCross);

    my $nextState = {
        leftSide => $nextLeftSide,
        rightSide => $nextRightSide,
        torchSide => $nextTorchSide,
        elapsedTimeMinutes => $nextElapsedTimeMinutes
    };

    return $nextState;
}

sub nextStateFromRight {
    my ($state, $personsToCross) = @_;

    my $leftSide = $state->{leftSide};
    my $nextLeftSide = nextLeftSideFromRight($leftSide, $personsToCross);
    my $rightSide = $state->{rightSide};
    my $nextRightSide = nextRightSideFromRight($rightSide, $personsToCross);
    my $torchSide = $state->{torchSide};
    my $nextTorchSide = flipTorchSide($torchSide);
    my $elapsedTimeMinutes = $state->{elapsedTimeMinutes};
    my $nextElapsedTimeMinutes = nextElapsedTimeMinutes($elapsedTimeMinutes, $personsToCross);

    my $nextState = {
        leftSide => $nextLeftSide,
        rightSide => $nextRightSide,
        torchSide => $nextTorchSide,
        elapsedTimeMinutes => $nextElapsedTimeMinutes
    };

    return $nextState;
}

sub nextLeftSideFromLeft {
    my ($leftSide, $personsToCross) = @_;

    my @nextLeftSide = grep {
        my $person = $_;

        my $personId = $person->{id};

        my $keep = 1;
        for my $personToCross (@$personsToCross) {
            my $personToCrossId = $personToCross->{id};
            if($personId eq $personToCrossId) {
                $keep = 0;
                last;
            }
        }

        $keep;
    } @$leftSide;

    return \@nextLeftSide;
}

sub nextLeftSideFromRight {
    my ($leftSide, $personsToCross) = @_;

    my @nextLeftSide = (@$leftSide, @$personsToCross);

    return \@nextLeftSide;
}

sub nextRightSideFromLeft {
    my ($rightSide, $personsToCross) = @_;

    my @nextRightSide = (@$rightSide, @$personsToCross);

    return \@nextRightSide;
}

sub nextRightSideFromRight {
    my ($rightSide, $personsToCross) = @_;

    my @nextRightSide = grep {
        my $person = $_;

        my $personId = $person->{id};

        my $keep = 1;
        for my $personToCross (@$personsToCross) {
            my $personToCrossId = $personToCross->{id};
            if($personId eq $personToCrossId) {
                $keep = 0;
                last;
            }
        }

        $keep
    } @$rightSide;

    return \@nextRightSide;
}

sub flipTorchSide {
    my $torchSide = shift;

    if('LEFT' eq $torchSide) {
        return 'RIGHT';
    } elsif('RIGHT' eq $torchSide) {
        return 'LEFT';
    } else {
        die "Unknown torchSide $torchSide";
    }
}

sub nextElapsedTimeMinutes {
    my ($elapsedTimeMinutes, $personsToCross) = @_;

    my @timesToCrossMinutes = map {$_->{timeToCrossMinutes}} @$personsToCross;
    my $maxTimeToCrossMinutes = max(@timesToCrossMinutes);

    my $nextElapsedTimeMinutes = $elapsedTimeMinutes + $maxTimeToCrossMinutes;
}

sub finalStates {
    my $states = shift;

    my $currentStates = $states;
    while(1) {
        my $allDone = isAllDone($currentStates);
        last if $allDone;

        my @nextStates = map {
            my $state = $_;

            my $nextPossibleStates = nextPossibleStates($state);

            @$nextPossibleStates;
        } @$currentStates;

        $currentStates = \@nextStates;
    }

    return $currentStates;
}

sub finalTimesToCrossMinutes {
    my $states = shift;

    my $finalStates = finalStates($states);

    my @finalTimesToCrossMinutes = map {$_->{elapsedTimeMinutes}} @$finalStates;

    return \@finalTimesToCrossMinutes;
}

sub minTime {
    my $states = shift;

    my $finalTimes = finalTimesToCrossMinutes($states);
    my $minTime = min(@$finalTimes);

    return $minTime;
}

sub isAllDone {
    my $states = shift;

    my $allDone = all(sub {
        my $state = shift;
        return isDone($state);
    }, @$states);

    return $allDone;
}

sub isDone {
    my $state = shift;

    my $leftSide = $state->{leftSide};
    my $leftSideIsEmpty = @$leftSide == 0;

    return $leftSideIsEmpty;
}

sub all {
    my $predicate = shift;
    my @array = @_;

    for my $element (@array) {
        my $predicateEvaluation = $predicate->($element);
        if(!$predicateEvaluation) {
            return 0;
        }
    }

    return 1;
}